source/common/http/http_server_properties_cache_impl.cc

```diff
@@ -296,6 +297,30 @@ HttpServerPropertiesCacheImpl::getOrCreateHttp3StatusTracker(const Origin& origi
   return *it->second.h3_status_tracker;
 }
 
+bool HttpServerPropertiesCacheImpl::isHttp3Broken(const Origin& origin) {
```

Comment:
Sorry, I misread the implementation of getOrCreateHttp3StatusTracker() above. setPropertiesImpl() actually update canonical suffix mapping with the new origin. So even without your fix in isHttp3Broken(), CanonicalSuffix test (with marking origin1 broken and checking origin2 broken) may still pass. But if you mark origin2 broken and assert origin1 broken, it will fail. (However, I'm not sure if this is the desired behavior or origin 1 and origin2 brokenness should be consistant.) @RyanTheOptimist

Reply:
I ran the test without flipping my flag and it failed.
In that case I think what happens is that we call getOrCreateHttp3StatusTracker() and fail to find a tracker. We create a default tracker with h3_status as pending. So h3 is not considered broken.

Reply:
Can you test the case where origin2 is marked broken and check origin1?

Reply:
Whether the flag is flipped or not, marking origin2 broken doesn't affect origin1. origin1 returns false for h3 brokenness. 

When the flag is false, origin1 is not considered broken because we don't check for canonial suffix. When the flag is true, origin1 is not broken because we don't create a new alt_protocol and thus not update the canonical suffix.

Reply:
> Whether the flag is flipped or not, marking origin2 broken doesn't affect origin1. origin1 returns false for h3 brokenness.
> 

I'm not sure if this is desired or not. If we add alternative protocols for x.example.com, and then check shouldAttemptHttp3() on y.example.com, it returns yes. And then if we mark y.example.com HTTP/3 broken, x.example.com is regarded not broken.

Reply:
I think in prod, if we are to mark `y.example.com` broken, we will actually [create](https://github.com/envoyproxy/envoy/blob/44c5b009567b322b7599f600342e6d4152e87ecd/source/common/http/conn_pool_grid.cc#L495) a status tracker and set the canonical suffix. This will prevent `z.example.com` from attempting QUIC, but it won't prevent `x.example.com` because we will find its exact record so we don't check canonical suffix. Should I change the logic to only lookup canonical suffix but not the exact record?

Reply:
> This will prevent z.example.com from attempting QUIC

I'm not sure about this, I would have thought that canonical suffix (i.e. example.com) is tracking x.example.com, and following z.example.com will still attempt QUIC. Would love to see a test for this.